include/seahorn/Expr/ExprOpBinder.hh

```diff
@@ -126,14 +128,86 @@ struct BINDER {
 };
 } // namespace bind
 
+namespace typeCheck {
+namespace bindType {
+
+static inline bool binderCheck(Expr exp, TypeCheckerHelper &helper,
+                               ExprVector &boundTypes) {
+  if (exp->arity() == 0)
+    return false;
+
+  // makes sure that all of the binders arguments are constants and store their
+  // types
+  for (auto b = exp->args_begin(), e = exp->args_end() - 1; b != e; b++) {
```

Comment:
I have not looked at all the supporting function, but this might be more complex than you realize. I think at this point, I'd rather skip this check completely.

I am not sure how your `getBoundVars` is implemented. It should take nested quantifiers into account. Not sure if it does.

Similarly, the top-down visitor should take nested quantifiers into account and maintain a stack with what quantifier is being processed. Not all type checking results can be cached across different quantifiers (this is a conjecture right now, I need to think of an example to be sure). 

Essentially, in a top-down traverse we must maintain the list of all variables that are currently bound and their types, then whenever a bound variable is processed, it can look up that its type is correct.

A yet another alternative is to type a BVar variable as a free function, so that a body of the quantifier has a functional type. Then the binder checks that it properly binds the right variables. 

But again, this is getting complicated, but exactly in the cases that we do not currently use.

So I propose to type-check that the body is well typed and ignore checking the binders.